core/corehttp/hostname.go

```diff
@@ -0,0 +1,301 @@
+								}
+							}
+						}
+
+						// Rewrite the path to not use subdomains
+						r.URL.Path = pathPrefix + r.URL.Path
+						// Serve path request
+						childMux.ServeHTTP(w, r)
+						return
+					}
+					// If not, resource does not exist on the subdomain gateway, return 404
+					http.NotFound(w, r)
+					return
+				}
+			}
+			// We don't have a known gateway. Fallback on DNSLink lookup
+
+			// Wildcard HTTP Host check:
+			// 1. is wildcard DNSLink enabled (Gateway.NoDNSLink=false)?
+			// 2. does Host header include a fully qualified domain name (FQDN)?
+			// 3. does DNSLink record exist in DNS?
+			if !cfg.Gateway.NoDNSLink && isDNSLinkRequest(r, n) {
+				// rewrite path and handle as DNSLink
+				r.URL.Path = "/ipns/" + stripPort(r.Host) + r.URL.Path
+				childMux.ServeHTTP(w, r)
+				return
+			}
+
+			// else, treat it as an old school gateway, I guess.
+			childMux.ServeHTTP(w, r)
+		})
+		return childMux, nil
+	}
+}
+
+// isDNSLinkRequest returns bool that indicates if request
+// should return data from content path listed in DNSLink record (if exists)
+func isDNSLinkRequest(r *http.Request, n *core.IpfsNode) bool {
+	ctx, cancel := context.WithCancel(n.Context())
+	defer cancel()
+
+	fqdn := stripPort(r.Host)
+	if len(fqdn) > 0 && isd.IsDomain(fqdn) {
+		// Confirm DNSLink was not disabled for the fqdn or globally
+		name := "/ipns/" + fqdn
+		// check if DNSLink exists
+		_, err := n.Namesys.Resolve(ctx, name, nsopts.Depth(1))
+		if err == nil || err == namesys.ErrResolveRecursion {
+			return true
+		}
+	}
+	return false
+}
+
+func isSubdomainNamespace(ns string) bool {
+	switch ns {
+	case "ipfs", "ipns", "p2p", "ipld", "api":
+		return true
+	default:
+		return false
+	}
+}
+
+// Parses Host header of a subdomain-based URL and returns it's components
+// Note: hostname is host + optional port
+func parseSubdomains(hostHeader string) (hostname, ns, rootID string, ok bool) {
+	parts := subdomainGatewayRegex.FindStringSubmatch(hostHeader)
+	if len(parts) < 4 || !isSubdomainNamespace(parts[2]) {
+		return "", "", "", false
+	}
+	return parts[3], parts[2], parts[1], true
+}
+
+// Converts a hostname/path to a subdomain-based URL, if applicable.
+func toSubdomainURL(hostname, path string, query string) (url string, ok bool) {
+	parts := strings.SplitN(path, "/", 4)
+
+	var ns, rootID, rest string
+	switch len(parts) {
+	case 4:
+		rest = parts[3]
+		fallthrough
+	case 3:
+		ns = parts[1]
+		rootID = parts[2]
+	default:
+		return "", false
+	}
+
+	if !isSubdomainNamespace(ns) {
+		return "", false
+	}
+
+	// add prefix if query is present
+	if query != "" {
+		query = "?" + query
+	}
+
+	if ns == "api" || ns == "p2p" {
+		// API and P2P proxy use the same paths on subdomains:
+		// api.hostname/api/.. and p2p.hostname/p2p/..
+		return fmt.Sprintf(
+			"http://%s.%s/%s/%s/%s%s",
+			ns,
+			hostname,
+			ns,
+			rootID,
+			rest,
+			query,
+		), true
+	}
+
+	if rootCid, err := cid.Decode(rootID); err == nil {
```

Comment:
Let's break this into different cases for different namespaces. That way, we can explicitly fail if, e.g., the user passes `/ipfs/invalid`.

Reply:
Unless we just want to redirect that anyways? Seems like a waste.

Reply:
I'd keep it simple, don't feel like we should care about content resolution at this stage.

All `toSubdomainURL` should do is to prepare rootID for use in redirect URL.
Additional validation should happen after redirect, when subdomain request is processed.

Reply:
This is going to behave strangely. If we fail to parse the CID, we're going to redirect the user to `qmfoobar.ipfs.dweb.link` (lowercase).

Also, don't we need to handle IPNS keys explicitly? Specifically, inline peer IDs?

Reply:
Ugh, I always forget about ed25519 PeerIDs :man_facepalming:  

Added PeerID→CID normalization + tests in cf8a678ad7758e8a9c5c9c0e78b00caebef3385c

Reply:
I'm still not comfortable returning `http://qmbadcid.ipfs.dweb.link/some/path`. At that point, we'll have erased the original CID so we can't even display a reasonable error to the user.

Reply:
Hm.. erasure of the original CID is no longer a problem: https://github.com/ipfs/go-ipfs/pull/6982 / df3c593201e17e93cb8da253c28e4c4639f02922 got merged into this PR.

We now resolve every content path before returning redirect because we need to include data in response body for curl. We return HTTP 301 redirect only if original response was HTTP 200, and 200 requires successful IPFS resolve. 

To illustrate, `http://dweb.link/ipfs/QmInvalidCID` will produce HTTP Error 400 with the original (case-sensitive) content path:

```
invalid ipfs path: invalid path "/ipfs/QmInvalidCID": invalid CID: selected encoding not supported
```

I've added sharness test in 4637ab8fa81c1679f8e92e2521ee0e0014345b20 to ensure the content path is always validated before redirect.